ch5/ex-5-13-tail.c

```diff
@@ -0,0 +1,225 @@
+		}
+	}
+	return OK;
+}
+
+void writelines(void)
+{
+	char *bptr;
+
+	if (*firstln <= cursor) {
+		for (bptr = *firstln; bptr < cursor; bptr++)
+			putchar(*bptr);
+	} else {
+		for (bptr = *firstln; bptr < bufflimit; bptr++)
+			putchar(*bptr);
+		for (bptr = buff; bptr < cursor; bptr++)
+			putchar(*bptr);
+	}
+}
+
+int init_buff(unsigned long size)
+{
+	buffsize = size;
+	buff = malloc(buffsize * sizeof(char));
+	if (buff == NULL) {
+		printf("error: Insufficient memory (allocating buff)\n");
+		return ERROR;
+	}
+	cursor = buff;
+	bufflimit = buff + buffsize;
+	return OK;
+}
+
+int init_lines(void)
+{
+	lines = malloc(tailsize * sizeof(char *));
+	if (lines == NULL) {
+		printf("error: Insufficient memory (allocating lines)\n");
+		return ERROR;
+	}
+	firstln = currln = lines;
+	*lines = buff;
+	lineslimit = lines + tailsize;
+	return OK;
+}
+
+int enlargebuff(void)
+{
+	unsigned long used;
+	char *bptr, **lptr;
+	ptrdiff_t offset;
+
+	/* copy info about current buffer before it's replaced */
+	char *pbuff = buff, *pcursor = cursor, *pbufflimit = bufflimit;
+	char *pfirstchr = *firstln;
+	ptrdiff_t pfirst2limit = pbufflimit - pfirstchr;
+
+	if (buffsize == maxbuffsize) {
+		printf("error: Cannot increase buffer size beyond %ld\n",
+		       buffsize);
+		return ERROR;
+	}
+
+	/* replace buff */
+	buffsize = buffsize > maxbuffsize / 4 ? maxbuffsize : buffsize * 4;
+	if (init_buff(buffsize) != OK) {
+		return ERROR;
+	}
+
+	/* copy data from previous buff to new buff */
+	if (pfirstchr < pcursor) {
+		for (bptr = pfirstchr; bptr < pcursor;)
+			*cursor++ = *bptr++;
+		used = pcursor - pfirstchr;
+	} else {
+		for (bptr = pfirstchr; bptr < pbufflimit;)
+			*cursor++ = *bptr++;
+		for (bptr = pbuff; bptr < pcursor;)
+			*cursor++ = *bptr++;
+		used = pfirst2limit + (pcursor - pbuff);
+	}
+
+	/* update line pointers to point to lines' new positions in new buff,
+	   may copy uninitialised values */
+	for (lptr = lines; lptr < lineslimit; lptr++) {
+		offset = *lptr >= pfirstchr ? *lptr - pfirstchr :
+					      pfirst2limit + (*lptr - pbuff);
+		*lptr = buff + offset;
+	}
+
+	/* set cursor for new buff */
+	cursor = buff + used;
+
+	free(pbuff);
+	return OK;
+}
+
+int settailsize(int argc, char **argv)
+{
+	char c, *ptr;
+	unsigned int d;
+	unsigned long n = 0;
+	unsigned long max = ULONG_MAX;
+
+	if (argc <= 1) {
+		tailsize = DEFAULT_TAILSIZE;
+	} else if (argc == 2 && *argv[1] == '-') {
+		ptr = argv[1] + 1;
+		if (*ptr == '\0') /* match behaviour of linux tail */
+			n = DEFAULT_TAILSIZE;
+		while ((c = *ptr++) != '\0') {
+			if (!isdigit(c)) {
+				printf("error: Expected decimal integer: %s\n",
+				       argv[1] + 1);
+				return ERROR;
+			}
+			if (n > (max / 10)) {
+				printf("error: Argument too large, max: %lu\n",
+				       max);
+				return ERROR;
+			}
+			n *= 10;
+			d = (c - '0');
+			if (d > max - n) {
+				printf("error: Argument too large, max: %lu\n",
+				       max);
+				return ERROR;
+			}
+			n += d;
+		}
```

Comment:
There is a standard library function `atoi` which does this too but I think this is ok.

Reply:
I had a couple of concerns about atoi:

- I think it's _only_ an int and I was going trying to go overboard on supporting "unreasonable" input with ulong.
- itoa might produce unexpected results if the string value were greater than the max for the type.